flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1561,11 +1561,27 @@ class ExprLowering {
       // function type cast can be inserted.
       auto symbolAttr = builder.getSymbolRefAttr(caller.getMangledName());
       if (callSiteType.getNumResults() != funcOpType.getNumResults() ||
-          callSiteType.getNumInputs() != funcOpType.getNumInputs())
+          callSiteType.getNumInputs() != funcOpType.getNumInputs()) {
+        // Do not emit a warning here because this can happen in legal program
+        // if the function is not defined here and it was first passed as an
+        // argument without any more information.
         funcPointer =
             builder.create<fir::AddrOfOp>(getLoc(), funcOpType, symbolAttr);
-      else
+      } else if (callSiteType.getResults() != funcOpType.getResults()) {
+        // Implicit interface result type mismatch are not standard Fortran,
+        // but some compilers are not complaining about it.
+        // The front-end is not protecting lowering from this currently. Support
+        // this with a discouraging warning.
+        mlir::emitWarning(getLoc(),
+                          "return type mismatches were never standard"
+                          " compliant and may lead to undefined behaviors.");
+        // Cast the actual function to the current caller implicit type because
+        // that is the behavior we would get if we could not see the definition.
```

Comment:
Are you actually adding a type cast here? That's not correct.

What are the advantages of a warning here? I see none. The code is deeply incorrect and the compiler shouldn't try to fix it in the middle of lowering.